File: sdks/python-ws/examples/error-handling.py
"""
Error handling example: Demonstrate error handling patterns
"""

import asyncio
from realm import RealmWebSocketClient
from realm.exceptions import RateLimitError, AuthenticationError


async def main():
    client = RealmWebSocketClient(url="ws://localhost:8080")
    
    try:
        await client.connect()

        # Example 1: Rate limit handling
        try:
            await client.generate({
                "prompt": "Hello",
                "max_tokens": 50,
            })
        except Exception as e:
            if hasattr(e, "code") and e.code == "RATE_LIMIT_EXCEEDED":
                retry_after = getattr(e, "retry_after", 60)
                print(f"Rate limited! Retry after: {retry_after} seconds")
                await asyncio.sleep(retry_after)
                # Retry...
            else:
                raise

        # Example 2: Authentication error
        try:
            invalid_client = RealmWebSocketClient(
                url="ws://localhost:8080",
                api_key="invalid-key",
            )
            await invalid_client.connect()
        except Exception as e:
            if "auth" in str(e).lower() or "unauthorized" in str(e).lower():
                print("Authentication failed - check your API key")
            await invalid_client.disconnect()

        # Example 3: Function not found
        try:
            await client.generate({
                "prompt": "Test",
                "max_tokens": 10,
            })
        except Exception as e:
            if hasattr(e, "code") and e.code == "FUNCTION_NOT_FOUND":
                print("Function not available on this server")
            else:
                print(f"Other error: {e}")

    except Exception as e:
        print(f"Unexpected error: {e}")
    finally:
        await client.disconnect()


if __name__ == "__main__":
    asyncio.run(main())

File: sdks/nodejs-ws/src/examples/error-handling.ts
/**
 * Error handling example: Demonstrate error handling patterns
 */

import { RealmWebSocketClient } from "../client";

async function main() {
  const client = new RealmWebSocketClient({
    url: "ws://localhost:8080",
    apiKey: process.env.REALM_API_KEY,
  });

  try {
    await client.connect();

    // Example 1: Rate limit handling
    try {
      await client.generate({
        prompt: "Hello",
        max_tokens: 50,
      });
    } catch (error: any) {
      if (error.code === "RATE_LIMIT_EXCEEDED") {
        console.log(`Rate limited! Retry after: ${error.retryAfter} seconds`);
        // Wait and retry
        await new Promise(resolve => setTimeout(resolve, error.retryAfter * 1000));
        // Retry...
      } else {
        throw error;
      }
    }

    // Example 2: Authentication error
    try {
      const invalidClient = new RealmWebSocketClient({
        url: "ws://localhost:8080",
        apiKey: "invalid-key",
      });
      await invalidClient.connect();
    } catch (error: any) {
      if (error.code === "UNAUTHORIZED" || error.message.includes("auth")) {
        console.log("Authentication failed - check your API key");
      }
    }

    // Example 3: Function not found
    try {
      // This would fail if the function doesn't exist
      await client.generate({
        prompt: "Test",
        max_tokens: 10,
      });
    } catch (error: any) {
      if (error.code === "FUNCTION_NOT_FOUND") {
        console.log("Function not available on this server");
      } else {
        console.log("Other error:", error.message);
      }
    }

  } catch (error) {
    console.error("Unexpected error:", error);
  } finally {
    client.disconnect();
  }
}

main().catch(console.error);

